File: lib/product/local_linker.ml
type target = { relative_path : string; unit_ : Frontend_intf.source_unit }

type resolution = {
  caller : string;
  dependency : Frontend_intf.dependency;
  target : target;
}

type local_intent = { caller : string; dependency : Frontend_intf.dependency }
type lookup = Not_local | Missing | Found of target | Invalid of string

let normalize path = Util.normalize_slashes path

let drop_leading_dot path =
  let rec loop value =
    if Util.starts_with ~prefix:"./" value then
      loop (String.sub value 2 (String.length value - 2))
    else value
  in
  loop path

let trim_trailing_slash path =
  let rec loop value =
    if String.length value > 1 && Util.ends_with ~suffix:"/" value then
      loop (String.sub value 0 (String.length value - 1))
    else value
  in
  loop path

let relative_to_root ~root path =
  let root = normalize root |> drop_leading_dot |> trim_trailing_slash
  and path = normalize path |> drop_leading_dot in
  if root = "" || root = "." then Some path
  else if path = root then Some ""
  else
    let prefix = root ^ "/" in
    if Util.starts_with ~prefix path then
      Some
        (String.sub path (String.length prefix)
           (String.length path - String.length prefix))
    else None

let normalize_relative path =
  let rec consume stack = function
    | [] -> Some (String.concat "/" (List.rev stack))
    | "" :: rest | "." :: rest -> consume stack rest
    | ".." :: rest -> (
        match stack with
        | _ :: parent -> consume parent rest
        | [] -> None)
    | component :: rest -> consume (component :: stack) rest
  in
  consume [] (String.split_on_char '/' (normalize path))

let dirname path =
  match String.rindex_opt path '/' with
  | None -> ""
  | Some index -> String.sub path 0 index

let has_yaml_extension path =
  let path = String.lowercase_ascii path in
  Util.ends_with ~suffix:".yml" path || Util.ends_with ~suffix:".yaml" path

let without_leading_slash path =
  let rec loop value =
    if Util.starts_with ~prefix:"/" value then
      loop (String.sub value 1 (String.length value - 1))
    else value
  in
  loop path

let azure_template_reference reference =
  match String.rindex_opt reference '@' with
  | None -> Some reference
  | Some index ->
      let alias =
        String.sub reference (index + 1) (String.length reference - index - 1)
      in
      if String.lowercase_ascii alias = "self" then
        Some (String.sub reference 0 index)
      else None

let local_candidates ~caller (dependency : Frontend_intf.dependency) =
  let reference = normalize dependency.reference in
  match (dependency.provider, dependency.kind) with
  | Ir.Github, Action
    when Util.starts_with ~prefix:"./" reference
         || Util.starts_with ~prefix:"../" reference ->
      let path = drop_leading_dot reference in
      if has_yaml_extension path then Some [ path ]
      else Some [ path ^ "/action.yml"; path ^ "/action.yaml" ]
  | Gitlab, Include
    when (not (Util.contains ~needle:"://" reference))
         && (not (String.contains reference '@'))
         && (has_yaml_extension reference
            || Util.starts_with ~prefix:"/" reference
            || Util.starts_with ~prefix:"./" reference
            || Util.starts_with ~prefix:"../" reference) ->
      Some [ without_leading_slash (drop_leading_dot reference) ]
  | Azure, Template -> (
      match azure_template_reference reference with
      | None -> None
      | Some path when Util.contains ~needle:"${{" path -> None
      | Some path ->
          let path = normalize path in
          if Util.starts_with ~prefix:"/" path then
            Some [ without_leading_slash path ]
          else
            let parent = dirname caller in
            Some [ (if parent = "" then path else parent ^ "/" ^ path) ])
  | _ -> None

let source_index ~root sources =
  sources
  |> List.filter_map (fun (unit_ : Frontend_intf.source_unit) ->
      Option.bind (relative_to_root ~root unit_.path) (fun relative ->
          Option.map (fun path -> (path, unit_)) (normalize_relative relative)))
  |> List.sort (fun (left, _) (right, _) -> String.compare left right)

let find_target ~index ~caller dependency =
  match local_candidates ~caller dependency with
  | None -> Not_local
  | Some candidates -> (
      let normalized =
        candidates
        |> List.fold_left
             (fun result candidate ->
               match (result, normalize_relative candidate) with
               | (Error _ as error), _ -> error
               | Ok _, None ->
                   Error
                     ("local dependency escapes the workspace: "
                    ^ dependency.Frontend_intf.reference)
               | Ok values, Some value -> Ok (value :: values))
             (Ok [])
      in
      match normalized with
      | Error message -> Invalid message
      | Ok candidates -> (
          let matches =
            candidates
            |> List.concat_map (fun candidate ->
                List.filter_map
                  (fun (path, unit_) ->
                    if path = candidate then
                      Some { relative_path = path; unit_ }
                    else None)
                  index)
            |> List.sort_uniq (fun left right ->
                String.compare left.relative_path right.relative_path)
          in
          match matches with
          | [] -> Missing
          | [ target ] -> Found target
          | _ ->
              Invalid
                ("local dependency is ambiguous: "
               ^ dependency.Frontend_intf.reference)))

let problem (dependency : Frontend_intf.dependency) code message =
  { Frontend_intf.code; message; span = dependency.Frontend_intf.span }

let compilation_key ~root (compilation : Frontend_intf.compilation) =
  let path =
    Option.bind
      (relative_to_root ~root compilation.graph.source)
      normalize_relative
    |> Option.value ~default:(normalize compilation.graph.source)
  in
  Ir.provider_name compilation.provider ^ "\000" ^ path

let target_key provider target =
  Ir.provider_name provider ^ "\000" ^ target.relative_path

let collect ~root ~index initial =
  let rec visit seen compiled resolutions local_intents errors = function
    | [] ->
        if errors = [] then
          Ok (List.rev compiled, List.rev resolutions, List.rev local_intents)
        else Error (List.rev errors)
    | (compilation : Frontend_intf.compilation) :: pending ->
        let caller =
          Option.bind
            (relative_to_root ~root compilation.graph.source)
            normalize_relative
          |> Option.value ~default:(normalize compilation.graph.source)
        in
        let pending, resolutions, local_intents, errors, seen =
          compilation.dependencies
          |> List.fold_left
               (fun (pending, resolutions, local_intents, errors, seen)
                    dependency ->
                 match find_target ~index ~caller dependency with
                 | Not_local ->
                     (pending, resolutions, local_intents, errors, seen)
                 | Missing ->
                     ( pending,
                       resolutions,
                       { caller; dependency } :: local_intents,
                       errors,
                       seen )
                 | Invalid message ->
                     ( pending,
                       resolutions,
                       local_intents,
                       problem dependency "LOCAL-DEPENDENCY" message :: errors,
                       seen )
                 | Found target -> (
                     let resolution = { caller; dependency; target } in
                     let local_intents =
                       { caller; dependency } :: local_intents
                     in
                     let key = target_key compilation.provider target in
                     if List.mem key seen then
                       ( pending,
                         resolution :: resolutions,
                         local_intents,
                         errors,
                         seen )
                     else
                       match
                         Frontend.compile_string ~provider:compilation.provider
                           ~path:target.unit_.path ~source:target.unit_.source
                           ()
                       with
                       | Error problems ->
                           ( pending,
                             resolution :: resolutions,
                             local_intents,
                             List.rev_append problems errors,
                             key :: seen )
                       | Ok target_compilation ->
                           ( target_compilation :: pending,
                             resolution :: resolutions,
                             local_intents,
                             errors,
                             key :: seen )))
               (pending, resolutions, local_intents, errors, seen)
        in
        visit seen (compilation :: compiled) resolutions local_intents errors
          pending
  in
  let seen = List.map (compilation_key ~root) initial in
  visit seen [] [] [] [] initial

let evidence operation value =
  Abstract_value.string_constant value ~trust:Abstract_value.Trusted
    ~secrecy:Abstract_value.Public
    ~provenance:[ { origin = "workspace source"; span = Span.none; operation } ]

let replace_attribute name value attributes =
  (name, value) :: List.remove_assoc name attributes

let matching_resolution caller (resolutions : resolution list)
    (dependency : Frontend_intf.dependency) =
  List.find_opt
    (fun (resolution : resolution) ->
      resolution.caller = caller
      && resolution.dependency.provider = dependency.provider
      && resolution.dependency.kind = dependency.kind
      && resolution.dependency.reference = dependency.reference)
    resolutions

let matching_local_intent caller (local_intents : local_intent list)
    (dependency : Frontend_intf.dependency) =
  List.exists
    (fun (intent : local_intent) ->
      intent.caller = caller
      && intent.dependency.provider = dependency.provider
      && intent.dependency.kind = dependency.kind
      && intent.dependency.reference = dependency.reference)
    local_intents

let call_matches reference (node : Ir.node) =
  node.name = reference || node.name = "child:" ^ reference

let apply_resolution ~root resolutions local_intents
    (compilation : Frontend_intf.compilation) =
  let caller =
    Option.bind
      (relative_to_root ~root compilation.graph.source)
      normalize_relative
    |> Option.value ~default:(normalize compilation.graph.source)
  in
  let resolved =
    compilation.dependencies
    |> List.filter_map (fun dependency ->
        Option.map
          (fun resolution -> (dependency, resolution))
          (matching_resolution caller resolutions dependency))
  in
  let dependencies =
    compilation.dependencies
    |> List.map (fun dependency ->
        match matching_resolution caller resolutions dependency with
        | None ->
            if matching_local_intent caller local_intents dependency then
              { dependency with mutability = Frontend_intf.Local }
            else dependency
        | Some resolution ->
            {
              dependency with
              mutability = Frontend_intf.Local;
              status =
                Locked
                  {
                    revision = "local:" ^ resolution.target.relative_path;
                    digest =
                      "sha256:"
                      ^ Sha256.digest_string resolution.target.unit_.source;
                  };
            })
  and nodes =
    compilation.graph.nodes
    |> List.map (fun (node : Ir.node) ->
        if node.kind <> Ir.Call then node
        else
          match
            List.find_opt
              (fun ((dependency : Frontend_intf.dependency), _) ->
                call_matches dependency.reference node)
              resolved
          with
          | None -> node
          | Some (_, resolution) ->
              let digest =
                "sha256:" ^ Sha256.digest_string resolution.target.unit_.source
              and revision = "local:" ^ resolution.target.relative_path in
              {
                node with
                attributes =
                  node.attributes
                  |> replace_attribute "dependency.source"
                       (evidence "local source" revision)
                  |> replace_attribute "dependency.revision"
                       (evidence "local revision" revision)
                  |> replace_attribute "dependency.digest"
                       (evidence "local digest" digest);
                unknown = None;
              })
  in
  {
    compilation with
    dependencies;
    graph = Ir.finalize { compilation.graph with nodes };
  }

let compare_compilation left right =
  match
    String.compare
      (Ir.provider_name left.Frontend_intf.provider)
      (Ir.provider_name right.Frontend_intf.provider)
  with
  | 0 -> String.compare left.graph.source right.graph.source
  | comparison -> comparison

let link ~root ~sources compilations =
  let index = source_index ~root sources in
  match collect ~root ~index compilations with
  | Error _ as error -> error
  | Ok (compilations, resolutions, local_intents) ->
      Ok
        (compilations
        |> List.map (apply_resolution ~root resolutions local_intents)
        |> List.sort compare_compilation)

File: src/product/local_linker.rs
use crate::domain::{AbstractValue, Node, NodeKind, Provenance, Provider, Secrecy, Trust};
use crate::foundation::{Budget, Span, content_digest, normalize_slashes};
use crate::frontend::{
    Compilation, Dependency, DependencyKind, DependencyStatus, Mutability, compile,
};
use std::collections::{BTreeMap, BTreeSet};

type ResolutionKey = (Provider, String, DependencyKind, String);

/// Recursively compile and content-address local dependencies without touching
/// the filesystem. Every candidate must already be present in `sources`.
///
/// # Errors
/// Rejects paths that escape the snapshot, ambiguous action metadata, and
/// malformed local documents.
pub fn link_local<S: AsRef<str>>(
    sources: &BTreeMap<String, S>,
    roots: Vec<Compilation>,
    budget: Budget,
) -> Result<Vec<Compilation>, Vec<String>> {
    let sources = normalized_sources(sources)?;
    let mut compilations = roots;
    let mut seen: BTreeSet<(Provider, String)> = compilations
        .iter()
        .map(|compilation| {
            (
                compilation.provider,
                normalize(compilation.graph.source_path()),
            )
        })
        .collect();
    let mut resolutions: BTreeMap<ResolutionKey, String> = BTreeMap::new();
    let mut local_intents = BTreeSet::new();
    let mut index = 0;
    while index < compilations.len() {
        let provider = compilations[index].provider;
        let caller = normalize(compilations[index].graph.source_path());
        let dependencies = compilations[index].dependencies.clone();
        for dependency in &dependencies {
            let Some(candidates) = local_candidates(&caller, dependency)? else {
                continue;
            };
            let key = resolution_key(provider, &caller, dependency);
            local_intents.insert(key.clone());
            let matches: Vec<_> = candidates
                .into_iter()
                .filter(|candidate| sources.contains_key(candidate))
                .collect();
            if matches.len() > 1 {
                return Err(vec![format!(
                    "local dependency is ambiguous: {}",
                    dependency.reference
                )]);
            }
            let Some(target) = matches.first() else {
                continue;
            };
            resolutions.insert(key, target.clone());
            let target_key = (provider, target.clone());
            if seen.insert(target_key) {
                let Some(source) = sources.get(target) else {
                    return Err(vec!["local source index changed during linking".to_owned()]);
                };
                let target_compilation =
                    compile(provider, target, source, budget).map_err(|problems| {
                        problems
                            .into_iter()
                            .map(|problem| format!("{}: {}", problem.code, problem.message))
                            .collect::<Vec<_>>()
                    })?;
                compilations.push(target_compilation);
            }
        }
        index = index.saturating_add(1);
    }
    for compilation in &mut compilations {
        apply_resolutions(compilation, &sources, &resolutions, &local_intents)?;
    }
    compilations.sort_by(|left, right| {
        left.provider
            .cmp(&right.provider)
            .then(left.graph.source_path().cmp(right.graph.source_path()))
    });
    Ok(compilations)
}

fn normalized_sources<S: AsRef<str>>(
    sources: &BTreeMap<String, S>,
) -> Result<BTreeMap<String, &str>, Vec<String>> {
    let mut output = BTreeMap::new();
    for (path, source) in sources {
        let normalized = normalize_relative(path)
            .ok_or_else(|| vec![format!("workspace source path escapes the root: {path}")])?;
        if output.insert(normalized.clone(), source.as_ref()).is_some() {
            return Err(vec![format!(
                "duplicate normalized source path: {normalized}"
            )]);
        }
    }
    Ok(output)
}

fn normalize(value: &str) -> String {
    normalize_slashes(value).trim_start_matches("./").to_owned()
}

fn normalize_relative(value: &str) -> Option<String> {
    let normalized = normalize_slashes(value);
    let mut stack = Vec::new();
    for component in normalized.split('/') {
        match component {
            "" | "." => {}
            ".." => {
                stack.pop()?;
            }
            value => stack.push(value),
        }
    }
    Some(stack.join("/"))
}

fn dirname(value: &str) -> &str {
    value.rsplit_once('/').map_or("", |(parent, _)| parent)
}

fn yaml_extension(value: &str) -> bool {
    std::path::Path::new(value)
        .extension()
        .is_some_and(|extension| {
            extension.eq_ignore_ascii_case("yml") || extension.eq_ignore_ascii_case("yaml")
        })
}

fn local_candidates(
    caller: &str,
    dependency: &Dependency,
) -> Result<Option<Vec<String>>, Vec<String>> {
    let reference = normalize_slashes(&dependency.reference);
    let raw = match (dependency.provider, dependency.kind) {
        (Provider::Github, DependencyKind::Action)
            if reference.starts_with("./") || reference.starts_with("../") =>
        {
            let value = reference.trim_start_matches("./");
            if yaml_extension(value) {
                Some(vec![value.to_owned()])
            } else {
                Some(vec![
                    format!("{value}/action.yml"),
                    format!("{value}/action.yaml"),
                ])
            }
        }
        (Provider::Gitlab, DependencyKind::Include)
            if !reference.contains("://")
                && !reference.contains('@')
                && (yaml_extension(&reference)
                    || reference.starts_with('/')
                    || reference.starts_with("./")
                    || reference.starts_with("../")) =>
        {
            Some(vec![reference.trim_start_matches('/').to_owned()])
        }
        (Provider::Azure, DependencyKind::Template) => {
            let template = match reference.rsplit_once('@') {
                Some((path, alias)) if alias.eq_ignore_ascii_case("self") => path,
                Some(_) => return Ok(None),
                None => reference.as_str(),
            };
            if template.contains("${{") {
                return Ok(None);
            }
            if template.starts_with('/') {
                Some(vec![template.trim_start_matches('/').to_owned()])
            } else {
                let parent = dirname(caller);
                Some(vec![if parent.is_empty() {
                    template.to_owned()
                } else {
                    format!("{parent}/{template}")
                }])
            }
        }
        _ => None,
    };
    raw.map(|values| {
        values
            .into_iter()
            .map(|candidate| {
                normalize_relative(&candidate).ok_or_else(|| {
                    vec![format!(
                        "local dependency escapes the workspace: {}",
                        dependency.reference
                    )]
                })
            })
            .collect()
    })
    .transpose()
}

fn resolution_key(provider: Provider, caller: &str, dependency: &Dependency) -> ResolutionKey {
    (
        provider,
        caller.to_owned(),
        dependency.kind,
        dependency.reference.clone(),
    )
}

fn local_evidence(operation: &str, value: &str) -> AbstractValue {
    AbstractValue::string_constant(
        value,
        Trust::Trusted,
        Secrecy::Public,
        vec![Provenance {
            origin: "workspace source".to_owned(),
            span: Span::default(),
            operation: operation.to_owned(),
        }],
    )
}

fn call_matches(node: &Node, reference: &str) -> bool {
    node.name == reference || node.name == format!("child:{reference}")
}

fn apply_resolutions<S: AsRef<str>>(
    compilation: &mut Compilation,
    sources: &BTreeMap<String, S>,
    resolutions: &BTreeMap<ResolutionKey, String>,
    local_intents: &BTreeSet<ResolutionKey>,
) -> Result<(), Vec<String>> {
    let caller = normalize(compilation.graph.source_path());
    for dependency in &mut compilation.dependencies {
        let key = resolution_key(compilation.provider, &caller, dependency);
        if let Some(target) = resolutions.get(&key) {
            let Some(source) = sources.get(target) else {
                return Err(vec!["local source index changed during linking".to_owned()]);
            };
            dependency.mutability = Mutability::Local;
            dependency.status = DependencyStatus::Locked {
                revision: format!("local:{target}"),
                digest: content_digest(source.as_ref()),
            };
        } else if local_intents.contains(&key) {
            dependency.mutability = Mutability::Local;
        }
    }
    for node in &mut compilation.graph.nodes {
        if node.kind != NodeKind::Call {
            continue;
        }
        let Some((_, target)) = resolutions
            .iter()
            .find(|((provider, owner, _, reference), _)| {
                *provider == compilation.provider
                    && owner == &caller
                    && call_matches(node, reference)
            })
        else {
            continue;
        };
        let Some(source) = sources.get(target) else {
            return Err(vec!["local source index changed during linking".to_owned()]);
        };
        let revision = format!("local:{target}");
        node.attributes.insert(
            "dependency.source".to_owned(),
            local_evidence("local source", &revision),
        );
        node.attributes.insert(
            "dependency.revision".to_owned(),
            local_evidence("local revision", &revision),
        );
        node.attributes.insert(
            "dependency.digest".to_owned(),
            local_evidence("local digest", &content_digest(source.as_ref())),
        );
        node.unknown = None;
    }
    compilation.graph.finalize();
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::frontend::DependencyLocator;

    fn dependency(provider: Provider, kind: DependencyKind, reference: &str) -> Dependency {
        Dependency::unresolved(
            provider,
            kind,
            reference,
            DependencyLocator::Direct,
            Span::default(),
        )
    }

    #[test]
    fn workspace_path_helpers_are_platform_neutral_and_fail_closed() {
        assert_eq!(normalize("./a\\b.yml"), "a/b.yml");
        assert_eq!(normalize("././a.yml"), "a.yml");
        assert_eq!(
            normalize_relative("./a//b/../c.yml"),
            Some("a/c.yml".to_owned())
        );
        assert_eq!(normalize_relative("a/../../escape.yml"), None);
        assert_eq!(normalize_relative("../escape.yml"), None);
        assert_eq!(dirname("a/b/file.yml"), "a/b");
        assert_eq!(dirname("file.yml"), "");
        for path in ["a.yml", "a.YML", "a.yaml", "a.YaMl"] {
            assert!(yaml_extension(path), "YAML extension {path:?}");
        }
        for path in ["a", "a.json", "a.yml.txt"] {
            assert!(!yaml_extension(path), "non-YAML extension {path:?}");
        }

        let sources = BTreeMap::from([
            ("a/./b.yml".to_owned(), "first".to_owned()),
            ("a/b.yml".to_owned(), "second".to_owned()),
        ]);
        assert!(normalized_sources(&sources).is_err());
        assert!(
            normalized_sources(&BTreeMap::from([(
                "../escape.yml".to_owned(),
                "source".to_owned(),
            )]))
            .is_err()
        );
    }

    #[test]
    fn provider_local_candidates_cover_every_accepted_and_remote_shape() {
        let github_directory = dependency(Provider::Github, DependencyKind::Action, "./action");
        assert_eq!(
            local_candidates(".github/workflows/ci.yml", &github_directory),
            Ok(Some(vec![
                "action/action.yml".to_owned(),
                "action/action.yaml".to_owned(),
            ]))
        );
        let github_file = dependency(Provider::Github, DependencyKind::Action, "./action.YML");
        assert_eq!(
            local_candidates(".github/workflows/ci.yml", &github_file),
            Ok(Some(vec!["action.YML".to_owned()]))
        );
        for reference in ["owner/action@main", "action", "https://example.test/action"] {
            assert_eq!(
                local_candidates(
                    ".github/workflows/ci.yml",
                    &dependency(Provider::Github, DependencyKind::Action, reference),
                ),
                Ok(None)
            );
        }
        assert!(
            local_candidates(
                ".github/workflows/ci.yml",
                &dependency(Provider::Github, DependencyKind::Action, "../../../escape"),
            )
            .is_err()
        );

        for (reference, expected) in [
            ("include.yml", "include.yml"),
            ("/root", "root"),
            ("./relative", "relative"),
        ] {
            assert_eq!(
                local_candidates(
                    ".gitlab-ci.yml",
                    &dependency(Provider::Gitlab, DependencyKind::Include, reference),
                ),
                Ok(Some(vec![expected.to_owned()]))
            );
        }
        for reference in [
            "plain",
            "https://example.test/include.yml",
            "component@example",
        ] {
            assert_eq!(
                local_candidates(
                    ".gitlab-ci.yml",
                    &dependency(Provider::Gitlab, DependencyKind::Include, reference),
                ),
                Ok(None)
            );
        }
        assert!(
            local_candidates(
                ".gitlab-ci.yml",
                &dependency(Provider::Gitlab, DependencyKind::Include, "../parent"),
            )
            .is_err()
        );

        let azure = |reference| {
            local_candidates(
                "pipelines/root.yml",
                &dependency(Provider::Azure, DependencyKind::Template, reference),
            )
        };
        assert_eq!(
            azure("steps.yml"),
            Ok(Some(vec!["pipelines/steps.yml".to_owned()]))
        );
        assert_eq!(
            azure("/shared/steps.yml"),
            Ok(Some(vec!["shared/steps.yml".to_owned()]))
        );
        assert_eq!(
            azure("steps.yml@SELF"),
            Ok(Some(vec!["pipelines/steps.yml".to_owned()]))
        );
        assert_eq!(azure("steps.yml@external"), Ok(None));
        assert_eq!(azure("${{ parameters.template }}"), Ok(None));
        assert_eq!(
            local_candidates(
                "pipelines/root.yml",
                &dependency(Provider::Azure, DependencyKind::Task, "steps.yml"),
            ),
            Ok(None)
        );
    }

    #[test]
    fn evidence_and_call_matching_do_not_cross_semantic_identity() {
        let evidence = local_evidence("local digest", "sha256:value");
        assert_eq!(evidence.trust, Trust::Trusted);
        assert_eq!(evidence.secrecy, Secrecy::Public);
        assert_eq!(evidence.constants(), Some(&["sha256:value".to_owned()][..]));
        assert_eq!(evidence.provenance.len(), 1);
        assert_eq!(evidence.provenance[0].origin, "workspace source");
        assert_eq!(evidence.provenance[0].operation, "local digest");

        let direct = Node::simple(
            Provider::Github,
            NodeKind::Call,
            "./action",
            crate::domain::Phase::Run,
            Span::default(),
        );
        let child = Node::simple(
            Provider::Gitlab,
            NodeKind::Call,
            "child:include.yml",
            crate::domain::Phase::Run,
            Span::default(),
        );
        assert!(call_matches(&direct, "./action"));
        assert!(call_matches(&child, "include.yml"));
        assert!(!call_matches(&direct, "include.yml"));
        assert!(!call_matches(&child, "./action"));
    }

    #[test]
    fn ambiguous_action_metadata_is_rejected_and_missing_local_source_stays_explicit() {
        let workflow = "on: push\njobs:\n  build:\n    steps:\n      - uses: ./action\n";
        let root = compile(
            Provider::Github,
            ".github/workflows/ci.yml",
            workflow,
            Budget::default(),
        )
        .expect("root workflow");
        let ambiguous = BTreeMap::from([
            (".github/workflows/ci.yml".to_owned(), workflow.to_owned()),
            (
                "action/action.yml".to_owned(),
                "name: first\nruns:\n  using: composite\n  steps: []\n".to_owned(),
            ),
            (
                "action/action.yaml".to_owned(),
                "name: second\nruns:\n  using: composite\n  steps: []\n".to_owned(),
            ),
        ]);
        assert!(link_local(&ambiguous, vec![root.clone()], Budget::default()).is_err());

        let linked = link_local(
            &BTreeMap::<String, String>::new(),
            vec![root],
            Budget::default(),
        )
        .expect("missing source remains unresolved");
        assert_eq!(linked.len(), 1);
        assert_eq!(linked[0].dependencies[0].mutability, Mutability::Local);
        assert!(matches!(
            linked[0].dependencies[0].status,
            DependencyStatus::Unresolved(_)
        ));
        let call = linked[0]
            .graph
            .nodes
            .iter()
            .find(|node| node.kind == NodeKind::Call)
            .expect("local call");
        assert!(call.unknown.is_some());
        assert!(!call.attributes.contains_key("dependency.digest"));
    }

    #[test]
    fn resolution_application_requires_both_provider_and_owner_identity() {
        let workflow = "on: push\njobs:\n  build:\n    steps:\n      - uses: ./action.yml\n";
        let original = compile(
            Provider::Github,
            ".github/workflows/ci.yml",
            workflow,
            Budget::default(),
        )
        .expect("root workflow");
        let sources = BTreeMap::from([(
            "action.yml".to_owned(),
            "name: action\nruns:\n  using: composite\n  steps: []\n".to_owned(),
        )]);
        let dependency = original.dependencies.first().expect("local dependency");
        let wrong_keys = [
            resolution_key(Provider::Gitlab, ".github/workflows/ci.yml", dependency),
            resolution_key(Provider::Github, "other/workflow.yml", dependency),
        ];
        for wrong_key in wrong_keys {
            let mut compilation = original.clone();
            apply_resolutions(
                &mut compilation,
                &sources,
                &BTreeMap::from([(wrong_key, "action.yml".to_owned())]),
                &BTreeSet::new(),
            )
            .expect("unrelated resolution is ignored");
            let call = compilation
                .graph
                .nodes
                .iter()
                .find(|node| node.kind == NodeKind::Call)
                .expect("call node");
            assert!(call.unknown.is_some());
            assert!(!call.attributes.contains_key("dependency.digest"));
        }
    }
}
